skopt/space.py

```diff
@@ -0,0 +1,458 @@
+              dimensions),
+            - an instance of a `Dimension` object (for `Real`, `Integer` or
+              `Categorical` dimensions), or
+            - as a list of categories (for `Categorical` dimensions).
+        """
+        _dimensions = []
+
+        for dim in dimensions:
+            if isinstance(dim, Dimension):
+                _dimensions.append(dim)
+
+            elif (len(dim) == 3 and
+                  isinstance(dim[0], numbers.Real) and
+                  isinstance(dim[2], str)):
+                _dimensions.append(Real(*dim))
+
+            elif len(dim) > 2 or isinstance(dim[0], str):
+                _dimensions.append(Categorical(*dim))
+
+            elif isinstance(dim[0], numbers.Integral):
+                _dimensions.append(Integer(*dim))
+
+            elif isinstance(dim[0], numbers.Real):
+                _dimensions.append(Real(*dim))
+
+            else:
+                raise ValueError("Invalid grid component (got %s)." % dim)
+
+        self.dimensions = _dimensions
+
+    def rvs(self, n_samples=1, random_state=None):
+        """Draw random samples.
+
+        Parameters
+        ----------
+        * `n_samples` [int, default=1]:
+            Number of samples to be drawn from the space.
+
+        * `random_state` [int, RandomState instance, or None (default)]:
+            Set random state to something other than None for reproducible
+            results.
+
+        Returns
+        -------
+        * `points`: [array-like, shape=(n_points, n_dims)]
+           Points sampled from the space.
+        """
+        rng = check_random_state(random_state)
+
+        # Draw
+        columns = []
+
+        for dim in self.dimensions:
+            if sp_version < (0, 16):
+                columns.append(dim.rvs(n_samples=n_samples))
+            else:
+                columns.append(dim.rvs(n_samples=n_samples, random_state=rng))
+
+        # Transpose
+        rows = []
+
+        for i in range(n_samples):
+            r = []
+            for j in range(self.n_dims):
+                r.append(columns[j][i])
+
+            rows.append(r)
+
+        return np.asarray(rows)
+
+    def transform(self, X):
+        """Transform samples from the original space into a warped space.
+
+        Note: this transformation is expected to be used to project samples
+              in a space where numerical optimization is suited.
+
+        Parameters
+        ----------
+        * `X` [array-like, shape=(n_samples, n_dims)]:
+            The samples to transform.
+
+        Returns
+        -------
+        * `Xt` [array-like, shape=(n_samples, transformed_n_dims)]
+            The transformed samples.
+        """
+        # Pack by dimension
+        columns = []
+        for dim in self.dimensions:
+            columns.append([])
+
+        for i in range(len(X)):
+            for j in range(self.n_dims):
+                columns[j].append(X[i][j])
```

Comment:
Do you think it is worthy to have a `dtype` proprty for each `Distribution` object? If we do that we can simply do:

```python
transformed_columns = []
for i, dim in enumerate(self.dimensions):
    curr_dtype = dim.dtype
    transformed_columns.append(dim.transform(dim.columns[:, i].astype(curr_dtype)).reshape(len(X), -1))

return np.hstack(transformed_columns)
```